feat: support dotnet lambda container builds

mrkdeng opened the pull request:
#### Which issue(s) does this change fix?
<!-- Use the format #<issue-number>, e.g. #42 -->
#4453

#### Why is this change necessary?
- This change adds support for .NET lambda container builds. It allows mounting source directory with write permissions when building in container because .NET builds write output to the source directory.
- In most cases on posix system, Docker Daemon is running as root. This change solves this problem by running docker as a non-root user.

#### How does it address the issue?
- Add a new sam build CLI argument `--mount-with [READ|WRITE]` that enables mounting source code directory with READ/WRITE permissions. Then run Docker as a non-root user (unless current user is root) if mount with write, which gives user the least privilege 
- .NET lambda container builds always need `--mount-with WRITE` to be specified. If building without explicit `--mount-with`, a confirmation dialog pops up and asks write permissions to source code directory
- This change is backward compatible and not breaking since container builds for .NET was not supported. For other runtimes, they mount source code as READ only by default, unless with intently using `--mount-with WRITE`

#### What side effects does this change have?
There is permission issue when running .NET image as non-root user [#7868](https://github.com/dotnet/core/issues/7868). This change relies on docker images to set environment variables such as `DOTNET_CLI_HOME`, `XDG_DATA_HOME`

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [X] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [X] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [X] Write/update unit tests
- [X] Write/update integration tests
- [X] Write/update functional tests if needed
- [X] `make pr` passes
- [X] `make update-reproducible-reqs` if dependencies were changed
- [X] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

Beau-Gosse-dev commented on samcli/local/docker/lambda_build_container.py:
```diff
@@ -104,8 +130,23 @@ def __init__(  # pylint: disable=too-many-locals
+        """
+        Make tmp dirs on host
+        """
+        if self._host_tmp_base_dir and not os.path.exists(self._host_tmp_base_dir):
```
So we won't create it if it already exists, but then we'll delete it afterwards? Can we not use a common path like /tmp/samcli ? A user might actually use that for something. Can we just use a GUID or something for the path? This is on the host machine, right? It doesn't have to match the path inside the container, right?

jfuss commented on samcli/local/docker/container.py:
```diff
@@ -135,15 +139,15 @@ def create(self):
 
         if self._host_dir:
-            LOG.info("Mounting %s as %s:ro,delegated inside runtime container", self._host_dir, self._working_dir)
+            mount_mode = "rw" if self._mount_with_write else "ro,delegated"
```
Should this still have `,delegated` on the `rw`?

jfuss commented on samcli/local/docker/container.py:
```diff
@@ -157,6 +161,16 @@ def create(self):
+                # skip if user id is 0 (root)
+                if user_id != 0:
+                    group_id = effective_user["group_id"]
+                    kwargs["user"] = f"{user_id}:{group_id}" if group_id else user_id
```
Can you explain what this is for? I am having a hard time understanding the implications here.

jfuss commented on samcli/local/docker/lambda_build_container.py:
```diff
@@ -42,7 +46,16 @@ def __init__(  # pylint: disable=too-many-locals
     ):
+        # tmp dirs on host that will be mounted to allow write permissions to `/tmp/samcli/` on container
+        # only when `mount_with_write` is True
+        if mount_with_write:
+            self._host_tmp_dir = os.path.join(source_dir, "tmp")
+            self._host_tmp_base_dir = os.path.join(source_dir, "tmp", "samcli")
+        else:
+            self._host_tmp_dir = None
+            self._host_tmp_base_dir = None
```
```
self._host_temp_dir = None
self._host_tmp_base_dir = None

if mount_with_write:
            self._host_tmp_dir = os.path.join(source_dir, "tmp")
            self._host_tmp_base_dir = os.path.join(source_dir, "tmp", "samcli")
```

Is a much cleaner way to write this

jfuss commented on samcli/local/docker/lambda_build_container.py:
```diff
@@ -249,21 +295,27 @@ def _convert_to_container_dirs(host_paths_to_convert, host_to_container_path_map
         """
-        return f"{LambdaBuildContainer._IMAGE_URI_PREFIX}-{runtime}:" + LambdaBuildContainer.get_image_tag(architecture)
+        # a list of build methods need to get image based on build method itself instead of runtime
+        get_image_from_build_method = ["dotnet7"]
```
Can we move this logic to before the calling this method? Seems weird to put this here and just override the runtime. Also hardcodes specific `dotnet7` things in "random" places. Can we make this more abstract to support similar things for future cases?

For example, If `build_method` is provided, we use that over the runtime directly?

mrkdeng replied:
Done. I changed the name to `specified_workflow` to be more consistent with `workflow_config.py`

jfuss commented on tests/testing_utils.py:
```diff
@@ -66,7 +66,7 @@ def run_command(command_list, cwd=None, env=None, timeout=TIMEOUT) -> CommandRes
 
 
-def run_command_with_input(command_list, stdin_input, timeout=TIMEOUT, cwd=None) -> CommandResult:
+def run_command_with_input(command_list, stdin_input, timeout=TIMEOUT, cwd=None, env=None) -> CommandResult:
```
`env` is unused here.

jfuss reviewed (changes requested):
Did a first pass with some feedback. 

Thank you for adding integration tests but please make sure you are also adding unit tests to cover your changes.

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -136,6 +136,15 @@
     "By default the functions and layers are built in sequence",
 )
+@click.option(
+    "--mount-with-write",
```
For boolean options, we usually provide both values. So we can rename this to; `--mount-with-write/--mount-with-read`? Default value would be `--mount-with-read` which will resolve as `mount_with_write=False` down below.

mndeveci commented on samcli/lib/build/app_builder.py:
```diff
@@ -946,6 +961,7 @@ def _build_function_on_container(
             # "/." is a Docker thing that instructions the copy command to download contents of the folder only
             result_dir_in_container = response["result"]["artifacts_dir"] + "/."
             container.copy(result_dir_in_container, artifacts_dir)
+            container.remove_host_tmp_dirs()
```
This should be called in finally block

mndeveci commented on samcli/local/docker/container.py:
```diff
@@ -16,7 +16,7 @@
 from .exceptions import ContainerNotStartableException
 
-from .utils import to_posix_path, find_free_port, NoFreePortsError
+from .utils import to_posix_path, find_free_port, NoFreePortsError, get_posix_effective_user
```
nit: Please use absolute imports

mndeveci commented on samcli/local/docker/utils.py:
```diff
@@ -154,3 +155,22 @@ def get_docker_platform(architecture: str) -> str:
+    :returns Dictionary containing user and group id, or None for non posix systems
+    """
+    if os.name.lower() == 'posix':
+        effective_user = {}
```
Can we use `@dataclass` instead of dictionary here? It will be easy to use in other places (no need to hardcode its keys for instance).

mrkdeng commented:
@Beau-Gosse-dev @jfuss @mndeveci thank you for reviewing this PR. Push two commits with one addressing comments, another one updating/adding unit tests

jfuss commented on samcli/lib/build/workflow_config.py:
```diff
@@ -258,6 +254,47 @@ def _key(c: CONFIG) -> str:
     return True, None
 
 
+def supports_specified_workflow(specified_workflow: str) -> bool:
```
I am having an hard time understand what this is actually for. Could you describe the use-case for this?

Note: This is probably just a gap in my understanding.

mrkdeng replied:
Regarding [comment](https://github.com/aws/aws-sam-cli/pull/4665#discussion_r1099260481) to keep `_get_image` clean. If this method returns True, the `specified_workflow` (`BuildMethod` in `.yaml`) will overwrite `runtime` to get docker image. One concern is, `dotnet7` is not the only one that has `specified_workflow`, but we know for sure it is safe because container builds for `dotnet` wasn't supported. For other runtimes consuming `provided.al2`, it's hard to say it wouldn't break. So I think this check is necessary before passing down `specified_workflow`. Lmk if there's a better way to do this.

jfuss commented on samcli/local/docker/lambda_build_container.py:
```diff
@@ -142,12 +162,14 @@ def _make_request(
                         "dependency_manager": dependency_manager,
                         "application_framework": application_framework,
                     },
+                    "base_dir": container_dirs["base_dir"],
```
This json is used to call AWS Lambda Builders in an RPC mode within the container. `base_dir`, `specified_workflow`, and `mount_with_write` are not apart of that: https://github.com/aws/aws-lambda-builders/blob/develop/aws_lambda_builders/__main__.py#L117

Was this meant to be added here?

mrkdeng replied:
Remove if that's the only consumer.

mrkdeng commented on samcli/lib/build/app_builder.py:
```diff
@@ -7,15 +7,14 @@
 from aws_lambda_builders.builder import LambdaBuilder
 from aws_lambda_builders.exceptions import LambdaBuilderError
-
+from samcli.commands.build.utils import prompt_user_to_enable_mount_with_write
```
Yes, we could do it in the `build_context`. Currently we don't prompt click until we know `_mount_with_write` is a must-have. I might be wrong - if move it to `command`, we will need to introduce the functionality in `workflow_config.get_workflow_config` to determine if `_mount_with_write` is needed. Doing so does move the UX part out of the `lib`, but not separate those components ideally.

jfuss replied:
I don't have anything against prompting but it shouldn't be done within the lib. If the customer is passing in "mount with write", we can check that in the `samcli/commands/build` module. It doesn't need to be tied to a specific workflow (that will just cause more issues in the long run).

jfuss replied:
I don't have anything against prompting but it shouldn't be done within the lib. If the customer is passing in "mount with write", we can check that in the samcli/commands/build module. It doesn't need to be tied to a specific workflow (that will just cause more issues in the long run).

mrkdeng replied:
Moved prompting to `samcli.build.build_context`. I import `samcli.lib.build.work_config` and assume it's ok since there are already imports from `samlic.lib.build` to `samcli.build.build_context`

mrkdeng commented on samcli/commands/build/command.py:
```diff
@@ -136,6 +136,16 @@
     "By default the functions and layers are built in sequence",
 )
+@click.option(
+    "--mount-with-write/--mount-with-read",
```
Regarding [comment](https://github.com/aws/aws-sam-cli/pull/4665#discussion_r1100488846). Lmk `--mount-with-read` or `--no-mount-with-write` which makes more sense.

jfuss replied:
I can sync with @mndeveci. I would rather this be an option with enum values. I find this being a flag confusing and will make `samconfig.toml` more confusing than it is today.

jfuss replied:
I can sync with @mndeveci. I would rather this be an option with enum values. I find this being a flag confusing and will make samconfig.toml more confusing than it is today.

lucashuy commented on tests/integration/buildcmd/test_build_cmd.py:
```diff
@@ -1038,10 +1038,158 @@ def test_with_dotnetcore(self, runtime, code_uri, mode, architecture="x86_64"):
             )
         self.verify_docker_container_cleanedup(runtime)
 
+    @parameterized.expand(
+        [
+            ("dotnetcore3.1", "Dotnetcore3.1", None),
+            ("dotnet6", "Dotnet6", None),
+            ("dotnetcore3.1", "Dotnetcore3.1", "debug"),
+            ("dotnet6", "Dotnet6", "debug"),
+            ("provided.al2", "Dotnet7", None),
+        ]
+    )
+    @pytest.mark.flaky(reruns=3)
+    def test_with_dotnetcore_in_container_mount_with_write_explicit(
```
Can we also add the Docker skip decorator to this test?
```python
@skipIf(SKIP_DOCKER_TESTS or SKIP_DOCKER_BUILD, SKIP_DOCKER_MESSAGE)
```

lucashuy commented on tests/integration/buildcmd/test_build_cmd.py:
```diff
@@ -1038,10 +1038,158 @@ def test_with_dotnetcore(self, runtime, code_uri, mode, architecture="x86_64"):
+        )
+
+        expected = "{'message': 'Hello World'}"
+        if not SKIP_DOCKER_TESTS:
+            self._verify_invoke_built_function(
+                self.built_template, self.FUNCTION_LOGICAL_ID, self._make_parameter_override_arg(overrides), expected
+            )
+        self.verify_docker_container_cleanedup(runtime)
```
Similarly, since this is already a Docker related test, we can remove this check and just call the methods.

lucashuy commented on tests/integration/buildcmd/test_build_cmd.py:
```diff
@@ -1038,10 +1038,158 @@ def test_with_dotnetcore(self, runtime, code_uri, mode, architecture="x86_64"):
+        ]
+    )
+    @pytest.mark.flaky(reruns=3)
+    def test_with_dotnetcore_in_container_mount_with_write_interactive(
```
Same feedback regarding skipping Docker test decorator and `if` statement check at the end of this test

mndeveci commented on samcli/lib/build/workflow_config.py:
```diff
@@ -245,11 +245,7 @@ def _key(c: CONFIG) -> str:
-        "Try building without the container. Most .NET Core functions will build "
-        "successfully.",
-    }
+    unsupported: Dict[str, str] = {}
```
Why we are keeping empty `unsupported` dict here? It was `dotnet` builders before, but I can't understand what are we setting it now.

sriram-mv replied:
This makes sense, let's remove if we don't need it.

mndeveci commented on samcli/local/docker/container.py:
```diff
@@ -157,6 +168,15 @@ def create(self):
+        # Skip if current user is root on posix systems or non-posix systems
+        _root_user_id = "0"
+        effective_user = EffectiveUser.get_current_effective_user().to_effective_user_str()
+        if self._mount_with_write and effective_user and effective_user != _root_user_id:
```
Can we add some debug logs here to investigate any possible issue in the future?

mndeveci commented on samcli/local/docker/container.py:
```diff
@@ -256,6 +276,12 @@ def delete(self):
+            if self._host_tmp_dir:
+                host_tmp_dir_path = pathlib.Path(self._host_tmp_dir)
+                if host_tmp_dir_path.exists():
+                    shutil.rmtree(self._host_tmp_dir)
```
Same, let's add some debug logging

mndeveci commented on samcli/local/docker/container.py:
```diff
@@ -277,6 +303,10 @@ def start(self, input_data=None):
 
+        # Make tmp dir on the host
+        if self._mount_with_write and self._host_tmp_dir and not os.path.exists(self._host_tmp_dir):
+            os.mkdir(self._host_tmp_dir)
```
Same, let's add some debug logging

mrkdeng commented on samcli/commands/build/build_context.py:
```diff
@@ -296,7 +300,6 @@ def run(self):
             BuildError,
             BuildInsideContainerError,
             UnsupportedBuilderLibraryVersionError,
-            ContainerBuildNotSupported,
```
`dotnet` is the only runtime that throws this exception

mndeveci commented on samcli/commands/build/utils.py:
```diff
@@ -1,13 +1,65 @@
+    specified_workflow = None
+    code_uri = None
+
+    if len(functions) > 0:
+        function = functions[0]
+        runtime = function.runtime
+        code_uri = function.codeuri
+        # get specified_workflow if metadata exists
+        metadata = function.metadata
+        specified_workflow = metadata.get("BuildMethod", None) if metadata else None
+    elif len(layers) > 0:
+        layer = layers[0]
+        specified_workflow = layer.build_method
+        code_uri = layer.codeuri
```
Why we are checking just one function and layer? Shouldn't this loop through all of them to see if any of them satisfy the need for the prompt?

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -275,6 +287,7 @@ def do_cli(  # pylint: disable=too-many-locals, too-many-statements
         aws_region=click_ctx.region,
         hook_name=hook_name,
         build_in_source=build_in_source,
+        mount_with=mount_with,
```
I would set the default (READ) here in case it is `None` to make the rest of the implementation easier (we don't need to check if it is `None` or any other value.)

mndeveci commented on samcli/lib/build/app_builder.py:
```diff
@@ -682,7 +682,7 @@ def _build_function(  # pylint: disable=R1710
                     global_image = self._build_images.get(None)
                     image = self._build_images.get(function_name, global_image)
-
+                    build_workflow = specified_workflow if supports_specified_workflow(specified_workflow) else None
```
Shouldn't we set this higher up at line 650?

mndeveci replied:
Then can we re-use `specified_workflow` name or something else? Since `build_workflow` is too generic comparing to where it is coming from.

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -136,6 +136,15 @@
+@click.option(
+    "--mount-with",
+    "-mw",
+    type=click.Choice(["READ", "WRITE"], case_sensitive=False),
```
I would recommend writing an enum and comparing its value rather than keeping constant string (`"read", "write"`) in the code.

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -136,6 +137,16 @@
+    type=click.Choice(MountMode.values(), case_sensitive=False),
+    default=MountMode.READ.value,
+    help="Optional. Specify mount mode for building functions/layers inside container. "
+    "If mount with write permissions, some files in source code directory may be changed/added by the build process. "
```
Might be a better help text? 🤔 
```suggestion
    "If it is mounted with write permissions, some files in source code directory may be changed/added by the build process. "
```

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -175,6 +186,7 @@ def cli(
     config_env: str,
     hook_name: Optional[str],
     skip_prepare_infra: bool,
+    mount_with="READ",
```
Still using string, let's use the enum

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -230,6 +243,7 @@ def do_cli(  # pylint: disable=too-many-locals, too-many-statements
     exclude: Optional[Tuple[str, ...]],
     hook_name: Optional[str],
     build_in_source: Optional[bool],
+    mount_with="READ",
```
Same as above, please use enum rather than hardcoded string

mndeveci commented on samcli/commands/build/utils.py:
```diff
@@ -0,0 +1,106 @@
+
+    for function in resources_to_build.functions:
+        code_uri = function.codeuri
+        if code_uri:
+            runtime = function.runtime
+            code_dir = str(pathlib.Path(base_dir, code_uri).resolve())
+            # get specified_workflow if metadata exists
+            metadata = function.metadata
+            specified_workflow = metadata.get("BuildMethod", None) if metadata else None
+            config = get_workflow_config(runtime, code_dir, base_dir, specified_workflow)
+            # at least one function needs mount with write, return with prompting
+            if config.must_mount_with_write_in_container:
+                return prompt(config, code_dir)
```
If you use early-exit in if statements, we can remove some indentation here to remove some complexity while reading this code (see: https://medium.com/swlh/return-early-pattern-3d18a41bba8)
```suggestion
        if not code_uri:
            continue
        runtime = function.runtime
        code_dir = str(pathlib.Path(base_dir, code_uri).resolve())
        # get specified_workflow if metadata exists
        metadata = function.metadata
        specified_workflow = metadata.get("BuildMethod", None) if metadata else None
        config = get_workflow_config(runtime, code_dir, base_dir, specified_workflow)
        # at least one function needs mount with write, return with prompting
        if not config.must_mount_with_write_in_container:
            continue
        return prompt(config, code_dir)
```

Same can be done with the other loop for layer below

mndeveci commented on samcli/lib/build/workflow_config.py:
```diff
@@ -222,40 +222,25 @@ def get_workflow_config(
 
-    def _key(c: CONFIG) -> str:
-        return str(c.language) + str(c.dependency_manager) + str(c.application_framework)
+    supported_specified_workflow = ["dotnet7"]
```
Isn't this return `False` for every other runtime? In that case, it looks like we might be breaking their build workflow

mrkdeng replied:
This wouldn't break others. The last parameter passed to `_build_function_on_container` and downstream is newly added. It allows builds such as `dotnet7` with `provided.al2` as runtime and a specified workflow to use the specified workflow to get docker image. This method returns `False` for others, so no specified workflow being passed down and they will use `runtime` to get docker image as previous.

mndeveci commented on samcli/commands/build/build_context.py:
```diff
@@ -78,6 +78,7 @@ def __init__(
         locate_layer_nested: bool = False,
         hook_name: Optional[str] = None,
         build_in_source: Optional[bool] = None,
+        mount_with=MountMode.READ.value,
```
```suggestion
        mount_with: MountMode = MountMode.READ
```

mrkdeng replied:
Click.Choice takes [str] only. Workaround: Convert str to enum in BuildContext.__init__() https://github.com/aws/aws-sam-cli/blob/280b16030ddee5e1b8ff1121f39e29892a53186e/samcli/commands/build/build_context.py#L178

mndeveci commented on samcli/commands/build/build_context.py:
```diff
@@ -235,6 +239,19 @@ def run(self):
+        # boolean value indicates if mount with write or not, defaults to READ ONLY
+        mount_with_write = False
+        if self._use_container:
+            if self._mount_with.lower() == MountMode.WRITE.value.lower():
```
```suggestion
            if self._mount_with == MountMode.WRITE:
```

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -136,6 +137,16 @@
+@click.option(
+    "--mount-with",
+    "-mw",
+    type=click.Choice(MountMode.values(), case_sensitive=False),
```
Please check how this is done in other places, example: https://github.com/aws/aws-sam-cli/blob/develop/samcli/commands/local/cli_common/options.py#L230

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -175,6 +186,7 @@ def cli(
     config_env: str,
     hook_name: Optional[str],
     skip_prepare_infra: bool,
+    mount_with=MountMode.READ.value,
```
Please add typing and  default value might be redundant since click sets this up automatically

mndeveci commented on samcli/commands/build/command.py:
```diff
@@ -230,6 +243,7 @@ def do_cli(  # pylint: disable=too-many-locals, too-many-statements
     exclude: Optional[Tuple[str, ...]],
     hook_name: Optional[str],
     build_in_source: Optional[bool],
+    mount_with=MountMode.READ.value,
```
Default might come from click setting above, so we can add typing and remove default from here as well

mndeveci commented on samcli/commands/build/utils.py:
```diff
@@ -0,0 +1,110 @@
+    READ = "READ"
+    WRITE = "WRITE"
+
+    @classmethod
+    def values(cls) -> List[str]:
+        """
+        A getter to retrieve the accepted value list for mount mode
+
+        Returns: List[str]
+            The accepted mount mode list
+        """
+        return [e.value for e in cls]
```
If we can use `_members_` for click option, then we might remove this function here as well

mndeveci commented on samcli/local/docker/effective_user.py:
```diff
@@ -0,0 +1,39 @@
+            # Return None for non-posix systems
+            return None
+
+        root_user_id = "0"
```
can we move this to a constant?

mndeveci commented on samcli/local/docker/lambda_build_container.py:
```diff
@@ -83,16 +88,25 @@ def __init__(  # pylint: disable=too-many-locals
+        host_tmp_dir = None
+        if mount_with_write:
+            # Mounting tmp dir on the host as ``/tmp/samcli`` on container, which gives current user write permissions
+            host_tmp_dir = os.path.join(source_dir, str(uuid4()))
```
Let's use utility functions to create tmp directory rather than creating one in the source folder. See example utility here: https://github.com/aws/aws-sam-cli/blob/3a7a4c39d62456b623d4183857c6e253e5236cef/samcli/lib/package/local_files_utils.py#L14

mrkdeng replied:
Instead of `source_dir`, now using `../.aws-sam/build/`

mndeveci reviewed (approved):
Thanks for addressing all the comments, looking forward to ship this feature soon! 🥳

jfuss reviewed (approved):
LGTM.

Thanks @mrkdeng for handling all the feedback along the way. I know that can be a lot and time consuming but I think we ended up in a good state as a result.
